presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -364,6 +412,243 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
             return new RemoteSourceNode(exchange.getId(), childrenIds, exchange.getOutputSymbols(), exchange.getOrderingScheme(), exchange.getType());
         }
 
+        private PlanNode createRemoteMaterializedExchange(ExchangeNode exchange, RewriteContext<FragmentProperties> context)
+        {
+            checkArgument(exchange.getType() == REPARTITION, "Unexpected exchange type: %s", exchange.getType());
+
+            PartitioningScheme partitioningScheme = exchange.getPartitioningScheme();
+            checkArgument(!partitioningScheme.getHashColumn().isPresent(), "precomputed hashes are not supported in materializing exchanges");
+
+            PartitioningHandle partitioningHandle = partitioningScheme.getPartitioning().getHandle();
+            ConnectorId connectorId = partitioningHandle.getConnectorId()
+                    .orElseThrow(() -> new IllegalArgumentException("Unsupported partitioning handle: " + partitioningHandle));
+
+            Map<Symbol, Expression> constantExpressions = new HashMap<>();
+            Partitioning partitioning = partitioningScheme.getPartitioning();
+            List<Symbol> partitioningSymbols = assignPartitioningSymbols(partitioning, constantExpressions);
+            Map<Symbol, ColumnMetadata> temporaryTableColumns = assignColumnNames(exchange.getOutputSymbols(), constantExpressions.keySet());
+            List<String> partitioningColumns = partitioningSymbols.stream()
+                    .map(symbol -> temporaryTableColumns.get(symbol).getName())
+                    .collect(toImmutableList());
+            PartitioningMetadata partitioningMetadata = new PartitioningMetadata(partitioningHandle, partitioningColumns);
+
+            TableHandle temporaryTable = metadata.createTemporaryTable(
+                    session,
+                    connectorId.getCatalogName(),
```

Comment:
hmm... I was thinking later we could also materialize a system partitioning (once we have native format)... Now I realized it might not be that straight..  But we don't need to think about it now :) . 

One thought is we can have a `exchange` connector whose `ConnectorPartitioning` is `SystemPartitioning`.. And it actually writes to , for example HDFS...

Reply:
Never mind, looks like `SystemPartitioning` has some special treatment in `LocalExecutionPlanner`: https://github.com/prestodb/presto/blob/3fce12e0248662bc10d0a21c55615ea963adafa8/presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java#L376-L382